File: frontend/src/types/labeling.ts
/**
 * Labeling Types
 *
 * TypeScript interfaces for independent semantic labeling of SAE features.
 * Matches backend Pydantic schemas in src/schemas/labeling.py and src/models/labeling_job.py
 *
 * Backend API Contract:
 * - POST /api/v1/labeling - Start labeling job for an extraction
 * - GET /api/v1/labeling/:id - Get labeling job status
 * - GET /api/v1/labeling - List all labeling jobs
 * - POST /api/v1/labeling/:id/cancel - Cancel labeling job
 * - DELETE /api/v1/labeling/:id - Delete labeling job (keeps labels intact)
 * - POST /api/v1/extractions/:id/label - Convenience endpoint to label extraction
 *
 * WebSocket Events:
 * - Channel: labeling/{labeling_job_id}/progress
 *   - labeling:started - Job started
 *   - labeling:progress - Progress update
 *   - labeling:completed - Labeling completed
 *   - labeling:failed - Labeling failed
 *
 * Status transitions: queued -> labeling -> completed/failed/cancelled
 */

/**
 * Labeling job status.
 * Matches backend LabelingStatus enum.
 */
export enum LabelingStatus {
  QUEUED = 'queued',
  LABELING = 'labeling',
  COMPLETED = 'completed',
  FAILED = 'failed',
  CANCELLED = 'cancelled',
}

/**
 * Labeling method.
 * Matches backend LabelingMethod enum.
 */
export enum LabelingMethod {
  OPENAI = 'openai',
  OPENAI_COMPATIBLE = 'openai_compatible',
  LOCAL = 'local',
  MANUAL = 'manual',
}

/**
 * Labeling configuration request.
 * Matches backend LabelingConfigRequest schema.
 */
export interface LabelingConfigRequest {
  /** Extraction job ID to label features from */
  extraction_job_id: string;

  /** Labeling method (openai, openai_compatible, local, manual) */
  labeling_method: LabelingMethod;

  /** OpenAI model to use (default: gpt-4o-mini) */
  openai_model?: string;

  /** OpenAI API key (optional, uses server default if not provided) */
  openai_api_key?: string;

  /** OpenAI-compatible endpoint (e.g., /ollama/v1 for local proxy) */
  openai_compatible_endpoint?: string;

  /** OpenAI-compatible model (e.g., llama3.2) */
  openai_compatible_model?: string;

  /** Local model to use (default: meta-llama/Llama-3.2-1B) */
  local_model?: string;

  /** Prompt template ID to use for labeling (optional, uses default if not specified) */
  prompt_template_id?: string;

  /** Filter special tokens (<s>, </s>, etc.) from token analysis (default: true) */
  filter_special?: boolean;

  /** Filter single character tokens from token analysis (default: true) */
  filter_single_char?: boolean;

  /** Filter pure punctuation tokens from token analysis (default: true) */
  filter_punctuation?: boolean;

  /** Filter pure numeric tokens from token analysis (default: true) */
  filter_numbers?: boolean;

  /** Filter word fragments (BPE subwords) from token analysis (default: true) */
  filter_fragments?: boolean;

  /** Filter common stop words from token analysis (default: false) */
  filter_stop_words?: boolean;

  /** Save API requests to /tmp/ for testing and debugging (default: false) */
  save_requests_for_testing?: boolean;

  /** Sample rate for saving API requests (0.0-1.0, default: 1.0) */
  save_requests_sample_rate?: number;

  /** Export format for saved API requests: 'postman' (Postman collection), 'curl' (cURL command), or 'both' (default: both) */
  export_format?: 'postman' | 'curl' | 'both';

  /** Maximum tokens in LLM response (50-8000, default: 300) */
  max_tokens?: number;

  /** Batch size for labeling (1-100, default: 10) */
  /**
   * Skip features that already carry a verdict.
   *
   * Defaults to false — a full relabel, which is what this endpoint has
   * always done. True fills only the gaps.
   */
  skip_adjudicated?: boolean;

  batch_size?: number;

  /** Number of activation examples per feature (10-50) */
  max_examples?: number;

  /** API request timeout in seconds (30-600, default: 120) */
  api_timeout?: number;

  /** Save poor quality labels for debugging (default: false) */
  save_poor_quality_labels?: boolean;

  /** Sample rate for saving poor quality labels (0.0-1.0, default: 1.0) */
  poor_quality_sample_rate?: number;
}

/**
 * Labeling job.
 * Matches backend LabelingJob model and LabelingStatusResponse schema.
 */
export interface LabelingJob {
  /** Labeling job ID (format: label_{extraction_id}_{timestamp}) */
  id: string;

  /** Extraction job ID being labeled */
  extraction_job_id: string;

  // Configuration
  /** Labeling method used */
  labeling_method: LabelingMethod;

  /** OpenAI model used (if applicable) */
  openai_model?: string | null;

  /** OpenAI-compatible model used (if applicable) */
  openai_compatible_model?: string | null;

  /**
   * OpenAI-compatible endpoint used (if applicable).
   *
   * Part of WHICH JUDGE ran. A resume that cannot reproduce it is not a
   * resume — it is a second, different experiment on the remainder. The
   * encrypted `openai_api_key` is deliberately never returned; the worker
   * resolves it from the job, app settings or the environment.
   */
  openai_compatible_endpoint?: string | null;

  /** API request timeout in seconds, also part of the judge's configuration. */
  api_timeout?: number | null;

  /** Local model used (if applicable) */
  local_model?: string | null;

  /** Prompt template ID used (if applicable) */
  prompt_template_id?: string | null;

  // Token filtering configuration
  /** Filter special tokens */
  filter_special: boolean;

  /** Filter single character tokens */
  filter_single_char: boolean;

  /** Filter pure punctuation */
  filter_punctuation: boolean;

  /** Filter pure numeric tokens */
  filter_numbers: boolean;

  /** Filter word fragments */
  filter_fragments: boolean;

  /** Filter common stop words */
  filter_stop_words: boolean;

  /** Save API requests for testing */
  save_requests_for_testing: boolean;

  /** Export format for saved API requests */
  export_format: string;

  /** Maximum tokens in LLM response */
  max_tokens: number;

  /** Save poor quality labels for debugging */
  save_poor_quality_labels: boolean;

  /** Sample rate for saving poor quality labels (0.0-1.0) */
  poor_quality_sample_rate: number;

  // Status and progress
  /** Current labeling status */
  status: LabelingStatus;

  /** Labeling progress (0-1) */
  progress: number;

  /** Number of features labeled so far */
  features_labeled: number;

  /** Total features to label */
  total_features?: number | null;

  // Results
  /** Error message if failed */
  error_message?: string | null;

  /** Labeling statistics */
  statistics?: LabelingStatistics | null;

  // Celery
  /** Celery task ID */
  celery_task_id?: string | null;

  // Timestamps
  /** Job creation timestamp */
  created_at: string;

  /** Last update timestamp */
  updated_at: string;

  /** Labeling completion timestamp */
  completed_at?: string | null;

  // Extraction context (populated by API enrichment)
  /** Model name (from training or external SAE) */
  model_name?: string | null;

  /** Layer index from extraction job */
  layer_index?: number | null;

  /** Hook type from extraction job (e.g., 'residual', 'mlp', 'attention') */
  hook_type?: string | null;

  /** External SAE name (if applicable) */
  sae_name?: string | null;
}

/**
 * Labeling job statistics.
 * Included in LabelingJob.statistics when complete.
 */
export interface LabelingStatistics {
  /** Total features processed */
  total_features: number;

  /** Successfully labeled features */
  successfully_labeled: number;

  /** Failed label generations */
  failed_labels: number;

  /** Average label length in characters */
  avg_label_length: number;

  /** Labeling duration in seconds */
  labeling_duration_seconds: number;

  /** Labeling method used */
  labeling_method: string;
}

/**
 * Paginated labeling list response.
 * Matches backend LabelingListResponse schema.
 */
export interface LabelingListResponse {
  /** List of labeling jobs */
  data: LabelingJob[];

  /** Pagination metadata */
  meta: {
    total: number;
    limit: number;
    offset: number;
  };
}

/**
 * Labeling list query parameters.
 */
export interface LabelingListParams {
  /** Filter by extraction job ID */
  extraction_job_id?: string;

  /** Maximum number of results (1-100) */
  limit?: number;

  /** Number of results to skip */
  offset?: number;
}

/**
 * WebSocket labeling progress event payload.
 */
export interface LabelingProgressEvent {
  labeling_job_id: string;
  features_labeled: number;
  total_features: number;
  progress: number;
  status: LabelingStatus;
}

/**
 * WebSocket labeling completed event payload.
 */
export interface LabelingCompletedEvent {
  labeling_job_id: string;
  statistics: LabelingStatistics;
}

/**
 * WebSocket labeling failed event payload.
 */
export interface LabelingFailedEvent {
  labeling_job_id: string;
  error_message: string;
}

/**
 * What still needs labeling in one extraction, and the batch a resume would take.
 *
 * Returned by `GET /labeling/{extraction_job_id}/coverage`. This is the answer
 * to "what is left?", a question that could not be asked before: the outcome of
 * a labeling attempt was not recorded anywhere, and a failure was written as a
 * fake label with `label_source` and `labeled_at` both set — so every existing
 * count read 16,824 failed features across the estate as finished work.
 */
/**
 * One normalised failure reason and what it cost.
 *
 * `reason` is the first colon-delimited segment of the stored error — the
 * exception class, or the whole fixed sentence — so a variable tail (host,
 * port, feature id) does not split one cause into thousands of singleton
 * groups. `(no reason recorded)` is the backfilled placeholder: on the L46
 * extraction that is 14,560 of them, and it is the most important row here.
 */
export interface LabelingFailureReason {
  reason: string;
  count: number;
  /** Real features carrying this reason, so a count can be checked, not believed. */
  sample_feature_ids: string[];
}

export interface LabelingCoverage {
  extraction_job_id: string;
  /** Every feature in the extraction. */
  total: number;
  /** Raw per-status counts: pending | in_progress | succeeded | failed | skipped. */
  by_status: Record<string, number>;
  /**
   * Settled: a verdict exists, or the feature was deliberately left alone.
   * Includes `uninterpretable` — the judge's honest "no coherent pattern",
   * which is a RESULT and must never be redone.
   */
  adjudicated: number;
  /** Still lacking a verdict, whether or not a resume will take it. */
  outstanding: number;
  /**
   * What a resume would ACTUALLY take — measured with the batch query's own
   * predicate, so the button and the job cannot disagree.
   */
  remaining: number;
  /**
   * Outstanding features that have used up their retries. Reported rather than
   * hidden: `outstanding` minus `exhausted` is `remaining`, and without this a
   * card showing "39 failed" and no Resume button is arithmetic an operator
   * cannot reconcile.
   */
  exhausted: number;
  /** Claimed by a running job right now. */
  in_progress: number;
  /** A status the backend has not been taught about. Reported, never absorbed. */
  unclassified: number;
  /**
   * Adjudicated by a DIFFERENT judge than the one asked about. `null` unless a
   * prompt fingerprint and model were supplied — reporting 0 without them would
   * read as "nothing is stale", which is not known.
   */
  stale: number | null;
  failures_without_a_recorded_reason: number;
  /** Set when some failures predate per-feature error capture. */
  caveat: string | null;
  /**
   * Why the failures failed, commonest first. Answers "will re-running these
   * work?" before spending 32 GPU-hours finding out empirically.
   */
  failure_reasons: LabelingFailureReason[];

  /** Ready to POST to `/labeling/panel` unchanged. */
  resume_feature_ids: string[];
}


/**
 * A multi-batch labeling resume.
 *
 * Finishing a 53k-feature extraction is ~27 batches of 2000 at ~8 s/feature —
 * about 59 GPU-hours. Each batch is its own Celery task, because a single task
 * looping over them would exceed the 10 h soft limit and strand its message for
 * the 12 h visibility timeout.
 */
export interface LabelingResumeSweep {
  id: string;
  extraction_job_id: string;
  /** running | completed | failed | cancelled */
  status: string;
  batches_done: number;
  max_batches: number;
  batch_size: number;
  /** Counted from what each batch WROTE, never from batch count. */
  features_labeled: number;
  features_failed: number;
  last_labeling_job_id: string | null;
  cancel_requested_at: string | null;
  error_message: string | null;
  created_at: string;
  completed_at: string | null;
}


/**
 * What a job's labeling endpoint can serve right now.
 *
 * Resume defaults to the judge a job used, which is right when comparing halves
 * of one run — and useless when that judge is gone, which is exactly when
 * someone needs to resume. This is what lets a resume name a different model.
 *
 * `features.label_model` and `label_prompt_fingerprint` are recorded per
 * feature, so an extraction labelled by two judges stays honest about which
 * produced what.
 */
export interface AvailableJudges {
  labeling_job_id: string;
  endpoint: string | null;
  /** What the job originally used. May no longer be served. */
  original_model: string | null;
  /** True when the original is still available — a plain resume works. */
  original_available: boolean;
  models: string[];
  reachable: boolean;
  detail: string | null;
}

File: backend/src/schemas/labeling.py
"""
Pydantic schemas for feature labeling.

This module defines request and response schemas for semantic labeling
of features extracted from SAE models.
"""

from datetime import datetime
from typing import Optional, Dict, Any, List, Literal
from pydantic import BaseModel, Field, ConfigDict, field_validator


class LabelingConfigRequest(BaseModel):
    """
    Request schema for feature labeling configuration.

    Validates labeling parameters before starting a feature labeling job.
    """
    model_config = ConfigDict(from_attributes=True)

    extraction_job_id: str = Field(
        description="ID of the extraction job whose features should be labeled"
    )

    labeling_method: str = Field(
        description="Feature labeling method: 'openai' (OpenAI API, fast), 'local' (local LLM, slower), 'manual' (user-provided labels)"
    )

    # OpenAI configuration
    openai_model: Optional[str] = Field(
        default="gpt-4o-mini",
        description="OpenAI model for labeling when labeling_method='openai'. Options: 'gpt-4o-mini', 'gpt-4o', 'gpt-4-turbo'"
    )
    openai_api_key: Optional[str] = Field(
        default=None,
        description="OpenAI API key for labeling when labeling_method='openai'. If not provided, uses environment variable OPENAI_API_KEY"
    )

    # OpenAI-compatible configuration
    openai_compatible_endpoint: Optional[str] = Field(
        default=None,
        description="OpenAI-compatible endpoint URL for labeling when labeling_method='openai_compatible'. Must include /v1 suffix. Example: 'http://ollama.mcslab.io/v1'"
    )
    openai_compatible_model: Optional[str] = Field(
        default=None,
        description="Model name for OpenAI-compatible endpoint when labeling_method='openai_compatible'. Example: 'llama3.2'"
    )

    # Local model configuration
    local_model: Optional[str] = Field(
        default="meta-llama/Llama-3.2-1B",
        description="Local model to use for labeling when labeling_method='local'. Example: 'meta-llama/Llama-3.2-1B', 'microsoft/Phi-3-mini-4k-instruct'"
    )

    # Prompt template configuration
    prompt_template_id: Optional[str] = Field(
        default=None,
        description="ID of the labeling prompt template to use. If not specified, uses the default template."
    )

    # Token filtering configuration
    filter_special: bool = Field(
        default=True,
        description="Filter special tokens (<s>, </s>, etc.) from token analysis"
    )
    filter_single_char: bool = Field(
        default=True,
        description="Filter single character tokens from token analysis"
    )
    filter_punctuation: bool = Field(
        default=True,
        description="Filter pure punctuation tokens from token analysis"
    )
    filter_numbers: bool = Field(
        default=True,
        description="Filter pure numeric tokens from token analysis"
    )
    filter_fragments: bool = Field(
        default=True,
        description="Filter word fragments (BPE subwords like 'tion', 'ing') from token analysis"
    )
    filter_stop_words: bool = Field(
        default=False,
        description="Filter common stop words (the, and, is, etc.) from token analysis"
    )

    # Debugging configuration
    save_requests_for_testing: bool = Field(
        default=False,
        description="Save API requests to /tmp/ for testing and debugging in Postman/cURL"
    )
    save_requests_sample_rate: float = Field(
        default=1.0,
        ge=0.0,
        le=1.0,
        description="Sample rate for saving all requests (0.0-1.0). Set to 1.0 to save all requests, or lower to sample (e.g., 0.1 saves ~10%)."
    )
    export_format: Literal["postman", "curl", "both"] = Field(
        default="both",
        description="Export format for saved API requests: 'postman' (Postman collection), 'curl' (cURL command), or 'both'"
    )
    save_poor_quality_labels: bool = Field(
        default=False,
        description="Save poor quality labels for debugging. Helps identify ineffective labels like 'uncategorized' or generic responses."
    )
    poor_quality_sample_rate: float = Field(
        default=1.0,
        ge=0.0,
        le=1.0,
        description="Sample rate for saving poor quality labels (0.0-1.0). Set to 1.0 to save all poor quality labels, or lower to sample (e.g., 0.5 saves ~50%)."
    )

    # Optional resource configuration
    batch_size: Optional[int] = Field(
        default=10,
        ge=1,
        le=100,
        description="Number of features to label in parallel (1-100)"
    )

    # Example configuration (overrides template default)
    max_examples: Optional[int] = Field(
        default=None,
        ge=10,
        le=50,
        description="Number of activation examples per feature (10-50). If not specified, uses template's default."
    )

    # API timeout configuration
    api_timeout: Optional[float] = Field(
        default=120.0,
        ge=30.0,
        le=600.0,
        description="API request timeout in seconds (30-600). Default: 120s. Increase for larger models or more examples."
    )

    @field_validator('labeling_method')
    @classmethod
    def validate_labeling_method(cls, v: str) -> str:
        """Validate labeling method is supported."""
        valid_methods = ['openai', 'openai_compatible', 'local', 'manual']
        if v not in valid_methods:
            raise ValueError(f"labeling_method must be one of {valid_methods}")
        return v

    @field_validator('openai_api_key')
    @classmethod
    def validate_openai_key(cls, v: Optional[str], info) -> Optional[str]:
        """Validate OpenAI API key is provided when using OpenAI method."""
        # This validation will be enhanced in the service layer to check env var
        return v


class LabelingStatusResponse(BaseModel):
    """
    Response schema for labeling job status.

    Returns current status and progress of a feature labeling job.
    """
    model_config = ConfigDict(from_attributes=True)

    id: str
    extraction_job_id: str
    labeling_method: str
    openai_model: Optional[str] = None
    local_model: Optional[str] = None
    prompt_template_id: Optional[str] = None
    filter_special: bool = True
    filter_single_char: bool = True
    filter_punctuation: bool = True
    filter_numbers: bool = True
    filter_fragments: bool = True
    filter_stop_words: bool = False
    save_requests_for_testing: bool = False
    save_requests_sample_rate: float = 1.0
    export_format: str = "both"
    save_poor_quality_labels: bool = False
    poor_quality_sample_rate: float = 1.0
    status: str
    progress: float
    features_labeled: int
    total_features: Optional[int] = None
    error_message: Optional[str] = None
    statistics: Optional[Dict[str, Any]] = None
    created_at: datetime
    updated_at: datetime
    completed_at: Optional[datetime] = None


class LabelingListResponse(BaseModel):
    """
    Response schema for list of labeling jobs.

    Returns paginated list of labeling jobs with metadata.
    """
    model_config = ConfigDict(from_attributes=True)

    data: List[LabelingStatusResponse]
    meta: Dict[str, Any] = Field(
        description="Metadata including total count, limit, offset"
    )


class LabelingCreateResponse(BaseModel):
    """
    Response schema for labeling job creation.

    Returns the ID of the newly created labeling job.
    """
    model_config = ConfigDict(from_attributes=True)

    labeling_job_id: str
    celery_task_id: str
    status: str
    message: str


class RelabelFeaturesRequest(BaseModel):
    """
    Request schema for re-labeling features.

    Allows re-labeling of features from an existing extraction with new configuration.
    """
    model_config = ConfigDict(from_attributes=True)

    labeling_method: str = Field(
        description="Feature labeling method: 'openai' (OpenAI API), 'local' (local LLM)"
    )
    openai_model: Optional[str] = Field(
        default="gpt-4o-mini",
        description="OpenAI model for labeling"
    )
    openai_api_key: Optional[str] = Field(
        default=None,
        description="OpenAI API key. If not provided, uses environment variable"
    )
    local_model: Optional[str] = Field(
        default="meta-llama/Llama-3.2-1B",
        description="Local model to use for labeling"
    )
    batch_size: Optional[int] = Field(
        default=10,
        ge=1,
        le=100,
        description="Number of features to label in parallel (1-100)"
    )

    @field_validator('labeling_method')
    @classmethod
    def validate_labeling_method(cls, v: str) -> str:
        """Validate labeling method is supported."""
        valid_methods = ['openai', 'local']
        if v not in valid_methods:
            raise ValueError(f"labeling_method must be one of {valid_methods}")
        return v
